### PR TITLE
Fix response issues when overlapping IDP API calls.

### PR DESCRIPTION
**Pull request by Wathsara**
### Proposed changes in this pull request

When overlapping update calls on the same IDP,  responses from REST API had issues because of invalid cache hits. This PR is to fix that issue.

### CODE REVIEW

**Review by emswbandara — APPROVED** [review]

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java [hunk 1]
```diff
@@ -278,6 +278,50 @@ public IdentityProvider getIdPByResourceId(String resourceId, int tenantId, Stri
         return identityProvider;
     }
 
+    /**
+     * Get the updated IDP details using the resource ID.
+     *
+     * @param resourceId   resource ID of the identity provider.
```
Comment (somindatommy) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java:284]: Formatting error

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java [hunk 2]
```diff
@@ -278,6 +278,50 @@ public IdentityProvider getIdPByResourceId(String resourceId, int tenantId, Stri
...
+     * @param tenantId     Tenant ID of the identity provider.
+     * @param tenantDomain Tenant domain of the identity provider.
+     * @return Updated identity provider with given resource ID.
+     * @throws IdentityProviderManagementException
```
Comment (somindatommy) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java:288]: Need to complete the method comment with exception details

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java [hunk 3]
```diff
@@ -278,6 +278,50 @@ public IdentityProvider getIdPByResourceId(String resourceId, int tenantId, Stri
...
+        IdentityProvider identityProvider;
+        IdPResourceIdCacheKey cacheKey = new IdPResourceIdCacheKey(resourceId);
+        IdPCacheEntry entry = idPCacheByResourceId.getValueFromCache(cacheKey, tenantDomain);
+
+        if (entry != null) {
+            /*
+            Before updating an IDP we clear the cache. Therefore, if we get a cache hit again at this point, it should
+            be a reason of some other process done on the same IDP. Hence, we need to clear the cache before proceed
+            in order to generate a correct response.
+             */
```
Comment (somindatommy) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java:302]: Following is the block comment format
```
/*
 *
 *
 */
```

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java [hunk 4]
```diff
@@ -278,6 +278,50 @@ public IdentityProvider getIdPByResourceId(String resourceId, int tenantId, Stri
...
+
+        identityProvider = idPMgtDAO.getIDPbyResourceId(null, resourceId, tenantId, tenantDomain);
+
+        if (identityProvider != null) {
+            addIdPCache(identityProvider, tenantDomain);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("No IDP found with resource ID: %s in DB", resourceId));
+            }
+        }
```
Comment (somindatommy) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java:320]: Better if we can use `If without else` here. Not nice to have a separate else block just for a debug log.

Comment (somindatommy) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java:320]: We can return the method if `if (identityProvider == null) {` with a log. WDYT?

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java [hunk 5]
```diff
@@ -278,6 +278,49 @@ public IdentityProvider getIdPByResourceId(String resourceId, int tenantId, Stri
...
     }
 
+    /**
+     * Get the updated IDP details using the resource ID.
+     *
+     * @param resourceId   Resource ID of the identity provider.
+     * @param tenantId     Tenant ID of the identity provider.
+     * @param tenantDomain Tenant domain of the identity provider.
+     * @return Updated identity provider with given resource ID.
+     * @throws IdentityProviderManagementException Error when getting the iIdentity Provider.
```
Comment (AnuradhaSK) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAO.java:288]:
```suggestion
     * @throws IdentityProviderManagementException Error when getting the Identity Provider.
```

**Review by mevan-karu — APPROVED** [review]

**Pull request merged**